### PR TITLE
Chain ranged export

### PR DESCRIPTION
**Pull request by frrist**
## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: chain/store/snapshot.go [hunk 1]
```diff
@@ -68,6 +97,295 @@ func (cs *ChainStore) Import(ctx context.Context, r io.Reader) (*types.TipSet, e
...
+
+const (
+	Block taskType = iota
+	Message
+	Receipt
+	State
+	Dag
+)
+
+func (s *walkScheduler) work(ctx context.Context, todo *walkTask, results chan *walkResult) error {
```
Comment (frrist) [inline chain/store/snapshot.go:260]: This method needs review

File: node/impl/full/chain.go [hunk 2]
```diff
@@ -587,6 +590,80 @@ func (m *ChainModule) ChainGetMessage(ctx context.Context, mc cid.Cid) (*types.M
 	return cm.VMMessage(), nil
 }
 
+func (a ChainAPI) ChainExportRangeInternal(ctx context.Context, head, tail types.TipSetKey, cfg *api.ChainExportConfig) error {
```
Comment (frrist) [inline node/impl/full/chain.go:593]: this file needs review

File: chain/store/snapshot.go [hunk 3]
```diff
@@ -24,6 +27,32 @@ func (cs *ChainStore) UnionStore() bstore.Blockstore {
...
+		Version: 1,
+	}
+
+	if err := car.WriteHeader(h, w); err != nil {
+		return xerrors.Errorf("failed to write car header: %s", err)
+	}
+
+	unionBs := cs.UnionStore()
+	return cs.WalkSnapshotRange(ctx, head, tail, messages, receipts, stateroots, workers, func(c cid.Cid) error {
+		blk, err := unionBs.Get(ctx, c)
```
Comment (frrist) [inline chain/store/snapshot.go:42]: frustrating that we request each CID twice.

File: chain/store/snapshot.go [hunk 4]
```diff
@@ -68,6 +101,300 @@ func (cs *ChainStore) Import(ctx context.Context, r io.Reader) (*types.TipSet, e
...
+	}
+
+	pw, ctx := newWalkScheduler(ctx, store, cfg, tasks...)
+	results := make(chan *walkResult)
+	pw.startScheduler(ctx)
+	pw.startWorkers(ctx, results)
+
+	done := make(chan struct{})
+	var cbErr error
+	go func() {
```
Comment (hsanjuan) [inline chain/store/snapshot.go:376]: potential goroutine leak if things error somewhere, as close(results) is down below.

File: chain/store/snapshot.go [hunk 5]
```diff
@@ -68,6 +101,300 @@ func (cs *ChainStore) Import(ctx context.Context, r io.Reader) (*types.TipSet, e
...
+			// Because the workers may have exited early (due to the context being canceled).
+			for range s.out {
+				s.taskWg.Done()
+			}
+			// Because the workers may have enqueued additional tasks.
+			for range s.in {
+				s.taskWg.Done()
+			}
+			// now, the waitgroup should be at 0, and the goroutine that was _waiting_ on it should have exited.
+			log.Infow("walkScheduler stopped")
```
Comment (hsanjuan) [inline chain/store/snapshot.go:192]: need to re-approach synchronization here. Seems last info message never happens. Things hang on ctrl-c.

**Review by hsanjuan — CHANGES_REQUESTED** [review]

**Comment by hsanjuan** [comment]
At the end of the export the following is printer:

```
      INFO    chainstore      store/snapshot.go:283   block export    {"height": "897000"}
2022-10-26T16:07:50.734Z        INFO    chainstore      store/snapshot.go:283   block export    {"height": "897000"}
2022-10-26T16:07:50.735Z        INFO    chainstore      store/snapshot.go:283   block export    {"height": "897000"}
2022-10-26T16:07:50.735Z        INFO    chainstore      store/snapshot.go:283   block export    {"height": "897000"}
2022-10-26T16:07:50.735Z        INFO    chainstore      store/snapshot.go:283   block export    {"height": "897000"}
....
```

where height goes down to 0. That is very confusing since we are only doing a range export. Are we missing a check to finish on the starting height?

File: chain/store/snapshot.go [hunk 6]
```diff
@@ -269,21 +279,21 @@ func (s *walkScheduler) work(ctx context.Context, todo *walkTask, results chan *
...
 		if err != nil {
 			return err
 		}
 		var b types.BlockHeader
 		if err := b.UnmarshalCBOR(bytes.NewBuffer(data.RawData())); err != nil {
-			return xerrors.Errorf("unmarshaling block header (cid=%s): %w", blk, err)
+			return xerrors.Errorf("unmarshalling block header (cid=%s): %w", blk, err)
 		}
-		if b.Height%1000 == 0 {
+		if b.Height%1_000 == 0 {
```
Comment (hsanjuan) [inline chain/store/snapshot.go:292]: hmmm....what does this mean?

Comment (frrist) [reply chain/store/snapshot.go:292]: emit a log message every 1000 blocks
the underscore is serving as a comma for readability https://go.dev/doc/go1.13#language

File: chain/store/snapshot.go (hunk 4 above)
Comment (frrist) [reply chain/store/snapshot.go:390]: results is now always closed: https://github.com/filecoin-project/lotus/pull/9192/files#diff-740d4bd1f10c65674bc1d0a54f1fc4192d8008b0c08179a300b42116ea0454a9R408

File: chain/store/snapshot.go [hunk 7]
```diff
@@ -24,6 +28,35 @@ func (cs *ChainStore) UnionStore() bstore.Blockstore {
 	return bstore.Union(cs.stateBlockstore, cs.chainBlockstore)
 }
 
+func (cs *ChainStore) ExportRange(ctx context.Context, head, tail *types.TipSet, messages, receipts, stateroots bool, workers int64, cacheSize int, w io.Writer) error {
```
[the remark this answers is not in the document]
Comment (frrist) [reply chain/store/snapshot.go:31]: Done

**Comment by frrist** [comment]
One outstanding issue with this PR is that it sometimes cannot be stopped with Ctrl+C

**Comment by jennijuju** [comment]
We need this before nv18 for stable snapshot service

File: api/api_full.go [hunk 8]
```diff
@@ -171,6 +171,10 @@ type FullNode interface {
 	// If oldmsgskip is set, messages from before the requested roots are also not included.
 	ChainExport(ctx context.Context, nroots abi.ChainEpoch, oldmsgskip bool, tsk types.TipSetKey) (<-chan []byte, error) //perm:read
 
+	ChainExportRange(ctx context.Context, head, tail types.TipSetKey, cfg *ChainExportConfig) (<-chan []byte, error) //perm:read
```
Comment (magik6k) [inline api/api_full.go:174]: https://github.com/filecoin-project/go-jsonrpc/pull/88 adds support for calling methods on the client.

A callback accepting bytes on the client would be better as that lets you have backpressure, and is arguably more robust in general.

Comment (magik6k) [reply api/api_full.go:174]: (see https://github.com/filecoin-project/lotus/pull/10027/files for an example of how it's integrated in lotus)

**Comment by hsanjuan** [comment]
I'm continuing this on #10145 . I'm leaving this branch untouched since we are actually using it in production now (well, a different rebase on v0.19.0).

**Pull request closed without merging**